[Pandi Preshita] iP

Preshita01 opened the pull request:
# Jarvis: A chatbot that keeps track of your tasks for you

> “Your mind is for having ideas, not holding them.” – David Allen

* Have you always wanted your own personal assistant? 
* Are you tired of manually keeping track of your ever-changing tasks?

#### Then Jarvis is perfect for you :wink:

Jarvis is an **ABSOLUTELY FREE** and easy-to-use chatbot that can manage your tasks for you. With just a few short and simple commands, you will be able to do the following:
- [X] Add a todo task
- [x] Add a task with a deadline
- [X] Add an event task with a specific start and end time
- [X] Mark tasks as completed
- [X] Delete tasks
- [X] Retrieve the tasks that are to be done today


Want to try out Jarvis now? Just follow these 3 simple steps:

1. Download the JAR file from [here](https://github.com/Preshita01/ip/releases/tag/A-Jar)
2. Open the file and add your tasks
3. Let Jarvis manage your tasks for you!


Some upcoming features also include:
- [ ] Retrieve tasks for a particular date
- [ ] Retrieve overdue tasks
- [ ] Edit the information of existing tasks

### For Java programmers
If you are a Java programmer, you can use Jarvis to practice your Java skills as well! Here is the `main` method:

```javascript
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
````
### So what are you waiting for? [Download Jarvis now!](https://github.com/Preshita01/ip/releases/tag/A-Jar)

Commit:
Add Gradle support

Preshita01 committed:
Add greet, echo and exit

Preshita01 committed:
Add ability to store input by users and display them back when required

Preshita01 committed:
Add Task.java

Preshita01 committed:
Add mark as done

Preshita01 committed:
Add Todo.java

Preshita01 committed:
Add Deadline.java

Preshita01 committed:
Add Event.java

Preshita01 committed:
Add ToDos, Events and Deadlines

Preshita01 committed:
Add indents

Preshita01 committed:
Add test cases

Preshita01 committed:
Add DukeException.java

Preshita01 committed:
Add invalid and incomplete instructions exceptions

Preshita01 committed:
Replace Task array with ArrayList<Task>

Preshita01 committed:
Add delete

Preshita01 committed:
Remove unwanted space in return statement for toString

Preshita01 committed:
Change chatbot name to Jarvis

Preshita01 committed:
Add delete test case

Preshita01 committed:
Rename to Jarvis.java

Preshita01 committed:
Rename to JarvisException.java

Preshita01 committed:
Update EXPECTED.TXT to contain the correct expected output

Preshita01 committed:
Rename file back to DukeException

Preshita01 committed:
Rename file back to Duke.java and edit some print statements

Preshita01 committed:
Update expected output based on change in print statements in Duke.java

Preshita01 committed:
Remove unnecessary space

Preshita01 committed:
Remove unnecessary space

Preshita01 committed:
Change 'by' to 'at' for event tasks

Preshita01 committed:
Change 'by' to 'at' for event tasks

Preshita01 committed:
Change 'by' to 'at' for event tasks

Preshita01 committed:
Restore Task.java to Level-6 version

Preshita01 committed:
Change 'by' to 'at' for event tasks

Preshita01 committed:
Add additional add, markAsDone and delete functions to indicate when overwrite tasks saved on hard disk

Preshita01 committed:
Save the tasks in the hard disk automatically whenever the task list changes

Preshita01 committed:
Store deadline as a LocalDateTime object in Deadline.java

Preshita01 committed:
Add new error messages for wrongly formatted deadline inputs

Preshita01 committed:
Store event time as a LocalDateTime object in Event.java

Preshita01 committed:
Add new error messages for wrongly formatted event times

Preshita01 committed:
Add getDeadline to retrieve the deadline as a LocalDateTime object in Deadline.java

Preshita01 committed:
Add getEventTime to retrieve the event time as a LocalDateTime object in Event.java

Preshita01 committed:
Add 'today' command to retrieve tasks for the current day

Preshita01 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

Preshita01 committed:
Add toPrintToFile() to differentiate between 2 types of date time stamps used

Preshita01 committed:
Add toPrintToFile() to differentiate between 2 types of date time stamps used

Preshita01 committed:
Add toPrintToFile() to differentiate between 2 types of date time stamps used

Preshita01 committed:
Add toPrintToFile() to differentiate between 2 types of date time stamps used

Preshita01 committed:
Add Parser.java to deal with making sense of the user command

Preshita01 committed:
Add Storage.java to deal with loading tasks from the file and saving tasks in the file

Preshita01 committed:
Add TaskList.java to contain the task list and other list functions

Preshita01 committed:
Add Ui.java to deal with interactions with the user

Preshita01 committed:
Refactor the code to extract out closely related code as classes

Preshita01 committed:
Refactor code into 'duke' package

Preshita01 committed:
Add DukeTest.java

Preshita01 committed:
Add getDescription() function in Task.java

Preshita01 committed:
Add parseFind() function for new 'find' command

Preshita01 committed:
Add 'find' command

Preshita01 committed:
Add documentation for all java methods

Preshita01 committed:
Tweak code to comply with a coding standard

Preshita01 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java

Preshita01 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Task.java

Preshita01 committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d' into branch-A-Gradle

Preshita01 committed:
Add Gradle to automate running of unit tests

Preshita01 committed:
Add CheckStyle to detect coding style violations.

Preshita01 committed:
Merge branch 'branch-A-CheckStyle'

Preshita01 committed:
Change toString() to toPrintToFile() in Task.java

Preshita01 committed:
Add Main.java to create a GUI using FXML

Preshita01 committed:
Add DialogBox.java

Preshita01 committed:
Add image for user

Preshita01 committed:
Add image for Jarvis

Preshita01 committed:
Add MainWindow.fxml

Preshita01 committed:
Add DialogBox.fxml

Preshita01 committed:
Add MainWindow.java as the controller class specified in MainWindow.fxml

Preshita01 committed:
Add Launcher.java

Preshita01 committed:
Add getResponse() method in Duke.java

Preshita01 committed:
Modify all methods to return strings in Parser.java

Preshita01 committed:
Add more dependencies to create GUI

Preshita01 committed:
Merge branch 'branch-Level-10'

Preshita01 committed:
Rectify Event.java to record both the start and end times of an event

Preshita01 committed:
Modify parseToday() to check that task ends before the end of the day

Preshita01 committed:
Add a greeting at the start

Preshita01 committed:
Add a greeting at the start

Preshita01 committed:
Add JavaDoc for initialize() in MainWindow.java

Preshita01 committed:
Remove unused import statement

Yukun99 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+    @Override
+    public String toPrintToFile() {
+        return "[D]" + super.toString() + " (by: " + by + ")";
+    }
```
Maybe `getSaveString` or `getSaveFormat` would be a better name for the function?

Yukun99 commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+
+    public static DialogBox getUserDialog(String text, Image img) {
+        return new DialogBox(text, img);
+    }
```
This looks more like a creation method, maybe `createUserDialog` might be better?

Yukun99 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,69 @@
+                    return Parser.parseFind(input);
+                } else {
+                    throw new DukeException("OOPS!!! I'm sorry, but I don't " +
+                            "know what that means :-(");
```
Might be better to create a constant for this so that you can edit all messages easily from the top of the class.

Yukun99 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,69 @@
+            try {
+                if (Parser.parseCommand(input).equals("list")) {
+                    if (TaskList.getCounter() == 0) {
+                        return "\tThere are currently no tasks on your list :)";
```
Might be better to create a constant for this so that you can edit all messages easily from the top of the class.

Yukun99 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,69 @@
+                return "" + e;
+            }
+        }
+        return "Bye! Hope to see you soon :)";
```
Might be better to create a constant for this so that you can edit all messages easily from the top of the class.

Yukun99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,267 @@
+        result += "Here are the matching tasks in your list:\n";
+        int num = 1;
+        search = search.substring(5);
+        for (int i = 0; i < TaskList.getTaskList().size(); i++) {
```
Would it perhaps be better to abstract getting size into a separate method so that other developers can't modify your task list improperly?

BuzzingBuzzard reviewed (commented):
Overall coding standard is strictly followed. One thing: Maybe In every javadoc comment there should be an empty line between description and parameter section? Otherwise good as far as I can see 👍

Yukun99 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,109 @@
+        return this.getStatusIcon() + this.description;
+    }
+
+    /**
+     * To write to the file in user's hard disk that stores a list of tasks or to overwrite
+     * the contents of this file
+     * @param filePath the relative path to the file
+     * @param textToAdd the content that is to be written
+     * @throws IOException if there is an error in writing to/overwriting the file
+     */
+    private static void writeToFile(String filePath, String textToAdd) throws IOException {
+        FileWriter fw = new FileWriter(filePath);
+        fw.write(textToAdd);
+        fw.close();
+    }
+
+    /**
+     * To append content to the file in user's hard disk that stores a list of tasks
+     * @param filePath the relative path to the file
+     * @param textToAppend the content that is to be appended
+     * @throws IOException if there is an error in appending to existing content of the file
+     */
+    private static void appendToFile(String filePath, String textToAppend) throws IOException {
+        FileWriter fw = new FileWriter(filePath, true); // create a FileWriter in append mode
+        fw.write(textToAppend);
+        fw.close();
+    }
```
would it be better to abstract these to a Storage class?

Yukun99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,27 @@
+     * input
+     */
+    public Ui() {
+        String logo = "\n" +
+                "       __       ___      .______     ____    ____  __       _______.\n" +
+                "      |  |     /   \\     |   _  \\    \\   \\  /   / |  |     /       |\n" +
+                "      |  |    /  ^  \\    |  |_)  |    \\   \\/   /  |  |    |   (----`\n" +
+                ".--.  |  |   /  /_\\  \\   |      /      \\      /   |  |     \\   \\    \n" +
+                "|  `--'  |  /  _____  \\  |  |\\  \\----.  \\    /    |  | .----)   |   \n" +
+                " \\______/  /__/     \\__\\ | _| `._____|   \\__/     |__| |_______/    \n" +
+                "                                                                    \n";
```
Might be better to declare this as a constant above?

Yukun99 reviewed (commented):
Not bad, but some methods can be abstracted and some constants can be moved to outside the functions for declaration for ease of editing and tracing.

Preshita01 committed:
Change 'preHeight' to 'maxHeight' to allow all of user's tasks to be displayed

Preshita01 committed:
Add more dependencies to pack JavaFX libraries into a generated JAR file

Preshita01 committed:
Add java asserts

Java asserts helps to check and document important assumptions that
should hold at various points in the code.

This is useful as programmers can be alerted if they are breaching
these assumptions elsewhere in the code.

Let's,
* include more java asserts for methods create task objects based on
the user input to check if the given details are valid/sensible

Preshita01 committed:
Code quality: structure code logically and add empty line between
description and parameter section in javadoc comments

The body of most methods is written as one block with no line breaks
in between to identify the various parts of the body that handle
different parts of the logic. There is also no empty line between the
description and parameter section in javadoc comments.

Grouping method bodies according to the logic they handle and adding
line breaks to separate the parts improves readability of the code.
Adding empty lines between the description and parameter section of the
javadoc comments also improve readability.

Let's,
* group method bodies according to the logic they handle and add line
breaks to separate the parts
* add empty lines between the description and parameter section of the
javadoc comments

Preshita01 committed:
Code quality: add empty line between description and parameter section
in javadoc comments

There is no empty line between the description and parameter section in
javadoc comments.

Adding empty lines between the description and parameter section of the
javadoc comments also improve readability.

Let's,
* add empty lines between the description and parameter section of the
javadoc comments

Preshita01 committed:
Code quality: structure code logically, add more guiding comments,
add empty line between description and parameter section in
javadoc comments

The body of most methods is written as one block with no line breaks
in between to identify the various parts of the body that handle
different parts of the logic. Comments are also minimal, which makes it
hard to easily understand the logic of the code.There is also no empty
line between the description and parameter section in javadoc comments.

Grouping method bodies according to the logic they handle and adding
line breaks to separate the parts improves readability of the code.
Adding more guiding comments and empty lines between the description and
parameter section of the javadoc comments also improve readability.

Let's,
* group method bodies according to the logic they handle and add line
breaks to separate the parts
* add more in-line comments for non-trivial lines of code
* add empty lines between the description and parameter section of the
javadoc comments

Preshita01 committed:
Code quality: add empty line between description and parameter section
in javadoc comments

There is no empty line between the description and parameter section in
javadoc comments.

Adding empty lines between the description and parameter section of the
javadoc comments also improve readability.

Let's,
* add empty lines between the description and parameter section of the
javadoc comments

Preshita01 committed:
Code quality: add empty line between description and parameter section
in javadoc comments

There is no empty line between the description and parameter section in
javadoc comments.

Adding empty lines between the description and parameter section of the
javadoc comments also improves readability.

Let's,
* add empty lines between the description and parameter section of the
javadoc comments

Preshita01 committed:
Code quality: structure code logically and store string outputs as
constants

The body of most methods is written as one block with no line breaks
in between to identify the various parts of the body that handle
different parts of the logic. Specific strings at different parts of the
code also makes it harder to understand why the string is returned
there, as well as to modify it.

Grouping method bodies according to the logic they handle and adding
line breaks to separate the parts improves readability of the code.
Storing the strings to be returned as aptly-named constants and then
calling on the constants themselves makes it easier to understand and
modify the strings.

Let's,
* group method bodies according to the logic they handle and add line
breaks to separate the parts
* add private constants outside of the methods and assign them to the
various string outputs

Preshita01 committed:
Code quality: structure code logically, store string outputs as
constants, add empty line between description and parameter section
in javadoc comments and add more guiding comments

The body of most methods is written as one block with no line breaks
in between to identify the various parts of the body that handle
different parts of the logic. Specific strings at different parts of the
code also makes it harder to understand why the string is returned
there, as well as to modify it. There is also no empty line between the
description and parameter section in javadoc comments. Comments are
also minimal, which makes it hard to easily understand the logic of the
code.

Grouping method bodies according to the logic they handle and adding
line breaks to separate the parts improves readability of the code.
Storing the strings to be returned as aptly-named constants and then
calling on the constants themselves makes it easier to understand and
modify the strings. Adding more guiding comments and empty lines between
the description and parameter section of the javadoc comments also
improve readability. Adding more guiding comments and empty lines
between the description and parameter section of the javadoc comments
also improve readability.

Let's,
* group method bodies according to the logic they handle and add line
breaks to separate the parts
* add private constants outside of the methods and assign them to the
various string outputs
* add empty lines between the description and parameter section of the
javadoc comments
* add more in-line comments for non-trivial lines of code

Preshita01 committed:
Add Note.java to store user notes

Preshita01 committed:
Add NoteList.java

Preshita01 committed:
Create a duke object using filepath to user tasks and notes

Preshita01 committed:
Note class: add parse methods for notes

Since the 'Note' class is a new addition to allow users to store their
notes using Jarvis, there are no methods to manipulate user inputs for
creating or deleting notes.

Parse methods for the 'Note' class are needed to create and delete note
objects.

Let's,
* add basic parse methods for notes, such as 'parseNote' and
'parseDeleteNote'

Preshita01 committed:
Modify constructor to retrieve user notes and modify 'getResponse()' to
recognise 'note', 'notes' and 'delete note' commands

Preshita01 committed:
Modify Storage constructor to create a file in user's hard disk to
store users' notes and modify 'retrieveFileContents()' to retrieve
these notes

Preshita01 committed:
Merge pull request #1 from Preshita01/branch-A-Assertions

Add java asserts

Preshita01 committed:
Merge branch 'master' of https://github.com/Preshita01/ip

Preshita01 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Deadline.java
#	src/main/java/duke/Event.java

Preshita01 committed:
Merge branch 'master' into branch-D-Notes

Preshita01 committed:
Merge pull request #2 from Preshita01/branch-A-CodeQuality

Improve code quality

Preshita01 committed:
Merge branch 'master' into branch-D-Notes

# Conflicts:
#	src/main/java/duke/Duke.java

Preshita01 committed:
Merge pull request #3 from Preshita01/branch-D-Notes

Add 'notes' functionality

Preshita01 committed:
Add GitHub Actions

Preshita01 committed:
Refactor code by shifting any output strings by Jarvis to Ui.java

Preshita01 committed:
Refactor code by shifting Jarvis' first greeting to the user to Ui.java

Preshita01 committed:
Refactor code by shifting any output strings by Jarvis to Ui.java and
any writings to the note file in user's hard disk to Storage.java

Preshita01 committed:
Refactor code by shifting any output strings by Jarvis to Ui.java and
any writings to the note file in user's hard disk to Storage.java

Preshita01 committed:
Refactor code by including all writings to the task and note files in
user's hard disk to Storage.java

Preshita01 committed:
Refactor code by shifting any writings to the task file in user's hard
disk to Storage.java

Preshita01 committed:
Refactor code by shifting any output strings by Jarvis to Ui.java and
any writings to the task file in user's hard disk to Storage.java

Preshita01 committed:
Refactor code by shifting any output strings by Jarvis to Ui.java

Preshita01 committed:
Change the header comments for javadoc comments for methods to start in
the form of 'returns', 'creates', etc.

Preshita01 committed:
Add javadoc comments and change the header comments of the javadoc
comments for methods such that they starts from 'returns', 'creates',
etc.

Preshita01 committed:
Add javadoc comments and shift any outputs by Jarvis to Ui.java

Preshita01 committed:
Add in-line comments and change the header comments of the javadoc
comments for methods such that they starts from 'returns', 'creates',
etc.

Preshita01 committed:
Add javadoc comments and in-line comments

Preshita01 committed:
Add javadoc comments and in-line comments

Preshita01 committed:
Format the header comments of the javadoc comments for the methods in
this class

Preshita01 committed:
Add javadoc comments and in-line comments

Preshita01 committed:
Add javadoc comments and in-line comments. Refactor code by shifting
any outputs by Jarvis to Ui.java.

Preshita01 committed:
Add javadoc and in-line comments. Refactor code by abstracting chunks
of a long method to improve readability.

Preshita01 committed:
Change the header comments of the javadoc comments for methods
such that they starts from 'returns', 'creates', etc.

Preshita01 committed:
Add javadoc comments and change the header comments of the javadoc
comments for methods such that they starts from 'returns', 'creates',
etc.

Preshita01 committed:
Add a space between the header comments and parameter description for
javadoc comments for the methods.

Preshita01 committed:
Add javadoc and in-line comments.

Preshita01 committed:
Change retrieveTaskFileContents() to retrieveNoteFileContents()

Preshita01 committed:
Rename DaDuke.png to DaJarvis.png

Preshita01 committed:
Add a space between package and import lines

Preshita01 committed:
Shift Jarvis' first greeting to Ui.java.

Preshita01 committed:
Format the tags correctly

Preshita01 committed:
Update to the correct expected output for the eventCreation test

Preshita01 committed:
Merge branch 'branch-A-Refactor'

Preshita01 committed:
Merge branch 'branch-A-CI'

Preshita01 committed:
Add a skyblue background for Jarvis and user responses and print error
messages in brown

Preshita01 committed:
Merge branch 'branch-A-BetterGui'

Preshita01 committed:
Adjust padding between dialog boxes

Preshita01 committed:
Add product screenshot

Preshita01 committed:
Add product screenshot

Preshita01 committed:
Set theme jekyll-theme-cayman

Preshita01 committed:
Add a User Guide

Preshita01 committed:
Set theme jekyll-theme-slate

Preshita01 committed:
Set theme jekyll-theme-time-machine

Preshita01 committed:
Set theme jekyll-theme-minimal

Preshita01 committed:
Set theme jekyll-theme-cayman

Preshita01 committed:
Add line breaks for improved readability

Preshita01 committed:
Modify listTasks() and listNotes() methods to have a header line before
printing out the lists

Preshita01 committed:
Add end() method to close the application

Preshita01 committed:
Add a call to Main.end() to close the application when user keys in the
command 'bye'

Preshita01 committed:
Update User Guide to include 'Quick Start' and 'Acknowledgements'

Preshita01 committed:
Update README.md to include the link to its JAR file

Preshita01 committed:
Remove unused code

Preshita01 committed:
Remove unnecessary tab for some Jarvis outputs

Preshita01 committed:
Rename 'duke' to 'jarvis' to match the name of the chatbot

Preshita01 committed:
Update the link to download the latest JAR file

Preshita01 committed:
Adjust top and bottom padding between dialog boxes

Preshita01 committed:
Update Ui.png to the screenshot of the latest Jarvis GUI

Preshita01 committed:
Update README.md to the latest version found in docs/README.md

Preshita01 committed:
Rename booleans to sound like booleans, such as from
'eventIsAfterDayStarts' to 'isEventAfterDayStarts'

Preshita01 committed:
Remove unnecessary space

Preshita01 committed:
Fix bug in the delete methods: Error message should be shown for wrongly
keyed in delete commands and negative task/note number

Preshita01 committed:
Fix bug for delete methods: Detect for 'delete note' command first
before 'delete'

Preshita01 committed:
Refactor error message for invalid task/note numbers so that it will
be highlighted in red in the GUI

Preshita01 committed:
Add additional error messages to handle more user errors

Preshita01 committed:
Modify error messages for parseEvent()

Preshita01 committed:
Add error handling for non-numerical and invalid date and time inputs
for deadline tasks

Preshita01 committed:
Add error handling for non-numerical and invalid date and time inputs
for event tasks

Preshita01 committed:
Add new error responses for Jarvis

Preshita01 committed:
Add handling of JarvisException

Preshita01 committed:
Modify method signatures to include JarvisException

Preshita01 committed:
Modify some method signatures to include JarvisException

Preshita01 committed:
Modify error handling message in parseEvent() to a more appropriate
message

Preshita01 committed:
Modify the conditions that trigger an 'empty description' error in
parseEvent()

Preshita01 committed:
Modify the conditions that trigger an 'empty description' error in
parseDeadline()

Preshita01 committed:
Add new error messages for Jarvis

Preshita01 committed:
Add javadoc and in-line comments

Preshita01 committed:
Modify the conditions that trigger an 'empty description' error in
parseTodo()

Preshita01 committed:
Handle possible user error where user keys in the description but not
the date and timings for a deadline task

Preshita01 committed:
Modify the conditions that trigger an 'incomplete done command' in
parseDone()

Preshita01 committed:
Add new error message

Preshita01 committed:
Add new error message

Preshita01 committed:
Merge branch 'branch-FixDoneCommandBugs'

Preshita01 committed:
Modify conditions to trigger 'missing search word/phrase' error in
findTasks()

Preshita01 committed:
Add new error message to be highlighted in red

Preshita01 committed:
Merge branch 'branch-FixFindCommandBugs'

Preshita01 committed:
Modify the conditions to trigger an 'incorrectly formatted note' error
in parseNote()

Preshita01 committed:
Add new error message

Preshita01 committed:
Add new error message to be highlighted in red

Preshita01 committed:
Merge branch 'branch-FixNoteCommandBugs'

Preshita01 committed:
Modify conditions to trigger 'incomplete command' error messages in
parseEvent() and parseDeadline()

Preshita01 committed:
Modify error messages

Preshita01 committed:
Modify error messages to be highlighted in red

Preshita01 committed:
Add a stage title

Preshita01 committed:
Update Ui.png

Preshita01 committed:
Update README.md to warns users that clashing tasks will not be
highlighted
